File: src/mk3/PythonPackage/src/tamaf/parallelbehavior.py
from .behavior import Behavior

class ParallelBehavior(Behavior):
    def __init__(self, agent, finishWhenAnyDone: bool = True):
        super().__init__(agent)
        self.subBehaviors: list[Behavior] = []
        self.finishWhenAnyDone = finishWhenAnyDone
        self.sharedData: dict = {}
    
    def AddSubBehavior(self, behavior: Behavior):
        behavior.agent = self.agent
        self.subBehaviors.append(behavior)

    def OnStart(self):
        for behavior in self.subBehaviors:
            behavior.OnStart()
    
    def Action(self):
        all_children_blocked = True
        
        for behavior in self.subBehaviors:
            if behavior.Done():
                continue
                
            if not behavior.IsBlocked():
                behavior.Execute()
                all_children_blocked = False

        if all_children_blocked:
            self.Block()

    def Done(self):
        if self.finishWhenAnyDone:
            for behavior in self.subBehaviors:
                if behavior.Done():
                    for b in self.subBehaviors:
                        b.OnEnd()
                    return True
            return False
        else:
            for behavior in self.subBehaviors:
                if not behavior.Done():
                    return False
            for behavior in self.subBehaviors:
                behavior.OnEnd()
            return True

    def OnEnd(self):
        self.sharedData = {}
        self.isStarted = False
        return 0

File: src/mk3/JavaPackage/src/main/java/tamaf/ParallelBehavior.java
package tamaf;

import java.util.ArrayList;
import java.util.List;

public class ParallelBehavior extends Behavior {
    private final List<Behavior> subBehaviors = new ArrayList<>();
    private boolean finishWhenAnyDone;
    protected java.util.Map<String, Object> sharedData = new java.util.HashMap<>();

    public ParallelBehavior(Agent agent, boolean finishWhenAnyDone) {
        super(agent);
        this.finishWhenAnyDone = finishWhenAnyDone;
    }
    
    public ParallelBehavior(boolean finishWhenAnyDone) {
        super();
        this.finishWhenAnyDone = finishWhenAnyDone;
    }

    public void addSubBehavior(Behavior b) {
        b.setAgent(this.agent);
        subBehaviors.add(b);
    }

    @Override
    public void onStart() {
        for (Behavior b : subBehaviors) {
            b.onStart();
        }
    }

    @Override
    public void action() {
        boolean allChildrenBlocked = true;
        for (Behavior b : subBehaviors) {
            if (b.done()) continue;

            if (!b.isBlocked()) {
                b.execute();
                allChildrenBlocked = false;
            }
        }

        if (allChildrenBlocked) {
            this.block();
        }
    }

    @Override
    public boolean done() {
        if (finishWhenAnyDone) {
            for (Behavior b : subBehaviors) {
                if (b.done()) {
                    for (Behavior sub : subBehaviors) sub.onEnd();
                    return true;
                }
            }
            return false;
        } else {
            for (Behavior b : subBehaviors) {
                if (!b.done()) return false;
            }
            for (Behavior b : subBehaviors) b.onEnd();
            return true;
        }
    }

    @Override
    public int onEnd() {
        this.sharedData.clear();
        this.setStarted(false);
        return 0;
    }
}
